### ISSUE TITLE
Disable direct manipulation for widget-bound selections

### ISSUE THREAD

**Issue by haldenl** [vega/vega-lite#4936]
~It seems like using two selections with one being a filter breaks selection logic. Clicking the scatter circle in the example below throws and error.~
__Even simpler: see Jeff's comment + example below.__

Downgrading to VL 2.* resolves the issue. Doesn't seem to be an issue without aggregation.

cc: @jheer 

Reproduced below:

[link to editor](https://vega.github.io/editor/#/url/vega-lite/N4KABGBEAmCGAutIC4zEgO1gWwKYqgGNYAnAZ0gF8AacKeE2DMgMwHsTsCBtOiUCIKgsAlgBt4uEgQFDBkMrjG5C8EWwwFIAZSUr4kPoMpGTEALq0IkbKQDWM+gE8ADvlSRCIkoWWRqUGQiAF7uYABMAAyRYDR0kLgYhGzQIhgA5jJGnmxiHFlyRBqpahoFhc5uWhhs2GmwYv5GQpCiStBaAPIkIulpTRWBeqrqmh6KyqqGhWbGVvIAHuUtsOnpJLjpCGE2uEwDcpDwrjsAjgCuTGqIagBu+PMtbWIdHgASHIoubADuUtNzbJOZbyVbrTbbLR4faPeTHKoeC5XEQ3ET3A5PETtLQAQUIhD0jFKmlMdDi1gm+lGIIUwwMqFkLXhOyCGT8sOsuGwLmOWgajQ5wixLwoqF4g0g3V6-WaYHMplhOjpIOszK0rPS7NlkAARmlXmhZdZbEtUABGACcAHYzYLjf1zdbInaoGkXOd6VBGBkHkbApIXAQzbLyYdntBRWBxRVIABNPbSWWWbVpFEq+Tx0hBp0h0kQUOQAAkZEIAAsuUgPKX4PAXGRkAB6Bv3LYAOj68FL5x1rfUDZL5dszYhAFoxCjcM2AMyt8Kts2tgBWZDK8xgCFgingkcZRFIke4u+sH3IuG+f2kyDNU+d2rxBOUROpV-CgrjCezVtvhUlPT6Y0gABVbQcQBWJHiPKATy+X5-ivAA2ABWN970JBBnzNBdkO1TNLzAS0bTfKV-xQICQLAyh5XzEBKCAA)

```json
{
  "data": {"name": "cars"},
  "transform": [
    {
      "filter": {
        "selection": "Select"
      }
    }
  ],
  "mark": { "type": "circle", "size": 200 },
  "encoding": {
    "color": {
      "condition": {
        "type": "nominal",
        "field": "Origin",
        "selection": "select"
      }
    },
    "x": {
      "aggregate": "mean",
      "type": "quantitative",
      "field": "Horsepower"
    },
    "y": {
      "aggregate": "mean",
      "type": "quantitative",
      "field": "Acceleration"
    }
  },
  "selection": {
    "select": {
      "type": "single",
      "empty": "all",
      "fields": [
        "Origin"
      ]
    },
    "Select": {
      "type": "single",
      "bind": {
        "max": 1971,
        "min": 1970,
        "input": "range",
        "step": 1
      },
      "fields": [
        "Year"
      ],
      "init": {
        "Year": 1970
      }
    }
  },
  "$schema": "https://vega.github.io/schema/vega-lite/v3.2.1.json",
  "datasets": {
    "cars": [{
      "Horsepower":130,
      "Acceleration":12,
      "Year": 1970,
      "Origin":"USA"
   },
   {
      "Horsepower":165,
      "Acceleration":11.5,
      "Year": 1971,
      "Origin":"USA"
   }]
  }
}
```

**Comment by jheer** [comment]
Here is [an online editor link](https://vega.github.io/editor/#/url/vega-lite/N4KABGBEAkDODGALApgWwIaQFxUQFzwAdYsB6UgN2QHN0A6agSz0QFcAjOxge1IRQyUa6ALQAbZskoBmOgCY6ARjoArWNwB2kADTgoAd0YATFtjByArAAZdESCkbV8Zyzb2Qj6PJhygIdinQxVmRYMwBtPX8wP2j-SABhblYNPAAnAE8zSAAhNPQKbjDbOLsAEWQJKkywAEVWIOYsnBErOmkADgsLADYrAHY5RTkS0sgK1G54fLxGeDqGiTxmsFa6Dp6LfulpOSse6Ssjo8VRuMgACWQglgywCQAzZDBkAA9CZHhvDXg7rzB2Iw0qYcBYAJx0fpg6HQuTSfqnKLnAAyjCeYGR6CMRmQaTMABY6D0etD8VtFCSztFIMjuNQwABxMoABTAHzSYHg6EIzB8YA6dGsFnxIosch6-Sp8QASjQeFocJAAKKwdTFJHUgCa1zxOD2+yldkBGnQmQA+oExGY2lCLNJFLswR1JRr4ohuTyNKFYGbUNdYKw0shsgBlKaMIJgAOEQjcYGQV0AXylsTGSRS6RWkGljGqYnQGiM6tK5Uqudxd3qjWWZjWnW6fUGw0NUAmUxmcwW1ZWaw2Wx2ewOxxOLcu1zEt3uaOebw+XwLvzA-0BwMQZnBkJhsPhiJLUFR6Mx2NxZh6RJJYLJCMprrstPpTNZ7M53N5ZkUbSFIrJ4pde+zcqaNkADqoR4LiRQ6LeUDaqaLhHD0o7GqaGQWkE1rrHIHRyNsiiKP0Fiju6MaMF6qq+v6gbBoqABiQbIEY3CoGAeDcGAGAANbPI8zxINwcyhAmpSJkiAC6ejJu46QFrADxxqgERIqm8QPIwE4nr40GQLAlSfLMQGKiGulfEJcQidE5lgKJJTacZ+kKjESKQEZYh6WYyl2MsHzZLApHUK5UFjMaRjuVppGEKweDZPkGjUMGo4YK88H4QlpHwQaWmwOBhDvkmUqQKplRFop-6wXirrWa6kCkcwoWlTq6U9EmSKWZJdgYGkHF1fEXnUVAsakVFZwFWprkhTg6QhMNvkAF59fqEk2cgPzcEYfndXY8DcGIcYbSpjBFdkspMEB+W9dkGhMaR6EtcNSWaWM52KgAjg0qS8rMVCBec6DUNQQa0OB2R+gW33UoVYjjfudKMiybK4i+PLeGD8QIEEfUefEjEYGlOCRHuECIdBEDDNB4nCbdTkrJjkBPVAr0FrM3iffFVW-f9whA4qINaPlENQzS04YliOJ4vlaMBQ9JYeEx6C42A+ME24BNgB+ZPNRZEkgImQA) to an even simpler version with a single selection that also breaks.

It appears the issue is that "normal" selection event logic is being generated alongside an input element binding. This leads to multiple errors:

1. In this case, the filter is well-defined. However, after filtering the data is aggregated and the field referenced by the selection no longer exists. This is why we see an error when a point is clicked.
2. The clearing logic (on background click or double-click) clears the selection value, which sets the slider value to an illegal value.

I believe the correct behavior is for an DOM-element-bound selection to NOT generate additional input event handling logic. This error is breaking the course work of multiple students so I consider it high priority!

**Comment by jheer** [comment]
It occurred to me that one way to get around this error might be to disable event handling for the selection. I tried setting `"on": null` but then I get compile errors about missing event specifications.

However, as a temporary workaround it looks like putting in the name of non-existent events is sufficient to resolve the bugs:

```json
"on": "none",
"clear": "none",
```

While definitely a hack, at least that gives us a workable solution!

@haldenl: this might be useful for students to know, too!

**Comment by arvind** [comment]
Thanks @haldenl and @jheer for including MVP specs. While the desired outcome is partially clear to me, how to best implement it is less so. Thoughts below:

> I believe the correct behavior is for a selection with a bind to a DOM element to NOT generate additional input event handling logic.

Not quite -- `bind` was designed to augment direct manipulation with query widgets, rather than have the latter replace the former. For instance, in our [interactive query widgets example](https://vega.github.io/editor/#/examples/vega-lite/interactive_query_widgets), (Year, Cylinder) pairs can be selected either by directly points or by manipulating the range sliders. 

This approach, however, breaks if the field a selection is projected over is no longer part of the dataset due to aggregation as in these examples. One option might be for the bind transform to disable direct manipulation if the projected fields do not participate in encoding. Of course, if no binding is specified, the selection will simply silently fail -- ideally we would issue a warning in such cases, but I'm not sure that we keep a record of which fields are available in a given unit (not just those participating in encoding but also those that may have been added in calculate transforms but are otherwise unused)? @domoritz? @kanitw? 

> The clearing logic (on background click or double-click) clears the selection value, which sets the slider value to an indeterminate value. This is particularly problematic when using multiple selections, as standard interactions defined for a separate selection (e.g., drawing a brush) may inadvertently reset the selection with the DOM element binding.

I'm not sure I understand the issue here or what the desired outcome is. If there are multiple selections, there is indeed a high likelihood that events will interfere (e.g., double clicking will clear out all active selections). To disambiguate, triggers for various selection functionality can be overridden as in @jheer's comment above or as in our [interactive splom example](https://vega.github.io/editor/#/examples/vega-lite/interactive_splom). Most functionality can also be disabled altogether using `false` (e.g., `clear: false`). Direct manipulation interaction currently can't, so I'd consider that a feature request albeit one that I'm hesitant to add as something users must explicitly set rather than something the compiler is able to infer.

**Comment by jheer** [comment]
Thanks @arvind! I downgraded the priority from P0 to P1 as we have the workaround described above, and changed the "bug" tag to "enhancement" as the issue stems from intentional design decisions, not errors.

My (inaccurate!) understanding this whole time was that the use of `bind` pre-empted normal direct manipulation interactions. (Maybe because I'm used to signals in Vega...) In any case, I've been laboring under this misconception for multiple years now, which is one point of evidence for it being unintuitive -- and our 512 students corroborate this confusion.

My inclination would be to have binding disable normal interactions by default. The combined use of bound widgets would then become opt-in. As a general rule of thumb, I prefer to have designers elect to use multiple functionalities rather than having them back out of them when they unexpectedly break -- especially when they break in what would otherwise be a completely reasonable use case.

The current default setup is brittle for a number of reasons:

1. Users might aggregate away fields defined in the selection, leading to dataflow execution errors.
2. The interactions can conflict with other selections, breaking one's interaction design.
3. The interactions can cause dynamic query widgets to enter an indeterminate state, which can be confusing and violate one's intended design. With menu/radio options, designers might not include `null` or intentionally wish to leave out some valid values in order to focus end users' attention. In the case of sliders, it is very common (and confusing) to see the slider suddenly go "null".

If we have consensus around a design change like this, we may not need explicit `false` or `null` settings for the `on` property. Thoughts?

**Comment by arvind** [comment]
Fixed with #5277.
